src/Stateless/Transition.cs

```diff
@@ -45,6 +48,11 @@ public Transition(TState source, TState destination, TTrigger trigger)
             public TTrigger Trigger { get { return _trigger; } }
 
             /// <summary>
+            /// Optional data associated with the transition.
+            /// </summary>
+            public object Tag { get { return _tag; } }
```

Comment:
No, neither do I. The 'Do()' is effectively redundant, as we already have an event that fires when the transition occurs, and the only thing missing is how to associate the two.
This is an incredibly common scenario, and has dozens of ways of coding. If you considered 'Tag' to be metadata then you might declare its type as `<int, Action<>>` and simply call the action in the central handler. 
Here's the OnOffExample modified to provide an ID and a function in the definition:
```
using System;
using Stateless;

namespace OnOffExample
{
    class TransitionMetadata
    {
        public int Id;
        public Action Activity;

        public TransitionMetadata(int id, Action activity)
        {
            this.Id = id;
            this.Activity = activity;
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            try
            {
                string on = "On", off = "Off";
                var space = ' ';

                var onOffSwitch = new StateMachine<string, char, TransitionMetadata>(off);
                onOffSwitch.OnTransitioned(t =>
                {
                    Console.WriteLine($"Detected transition ID {t.Tag.Id}");
                    t.Tag.Activity?.Invoke();
                });
                onOffSwitch.Configure(off).Permit(space, on).Tag(new TransitionMetadata(100, () => Console.WriteLine("On was transitioned!")));
                onOffSwitch.Configure(on).Permit(space, off).Tag(new TransitionMetadata(200, () => Console.WriteLine("Off was transitioned!")));

                Console.WriteLine("Press <space> to toggle the switch. Any other key will raise an error.");

                while (true)
                {
                    Console.WriteLine("Switch is in state: " + onOffSwitch.State);
                    var pressed = Console.ReadKey(true).KeyChar;
                    onOffSwitch.Fire(pressed);
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("Exception: " + ex.Message);
                Console.WriteLine("Press any key to continue...");
                Console.ReadKey(true);
            }
        }
    }
}
```

Sure, it's just another way of doing things, but it's a single point of execution for transitions, which I think is much cleaner. It also removes ambiguity between the event handler and the Do() action, so you define the order yourself.
Personally, I don't like code in my workflows. Ideally I like them to be serialisable, which means no embedded lambdas or object references, and when we are auto-generating workflows from users' definitions I would rather add those to a switch statement than embed them as lamdas throughout the StateMachine configuration section.